### PR TITLE
[11.0][ADD]purchase_product_usage

### PR DESCRIPTION
**Pull request by AaronHForgeFlow**
### Description
When purchasing, companies may need to specify in the different GL accounts depending on the usage the company is going to give to the part. Usages can be very diverse as for example: internal use, marketing use, testing use. So a product "Paper" could potentially be a marketing expense, or an office expense.

With this module, we can define different usages in the company for products. Purchase Users can specify the use the company expect to use the product for, and use the specific account for that.

### Usage
In order to test it, first activate usages for purchased products on settings.

Then, in the PO line set the usage of the purchased product. If the product is a consumable or a service and the valuation is periodic then the invoice will take the account defined in the usage. In any other case the normal behavior is expected.

### CODE REVIEW

**Review by HviorForgeFlow — CHANGES_REQUESTED** [review]
Functional review @aheficent 
- Purchase users are unable to select Usage on Quotations.
- Once I selected Product Usages in Purchases on Purchase Settings I can't remove Usages on products for Purchases options in user settings.

**Comment by AaronHForgeFlow** [comment]
Thanks for reporting @hveficent I think it is fixed now

File: purchase_product_usage/views/purchase_order_line_view.xml [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+        <field name="model">purchase.order</field>
+        <field name="inherit_id" ref="purchase.purchase_order_form"/>
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='order_line']/tree/field[@name='name']"
+                   position="after">
+                <field name="usage_id" groups="purchase_product_usage.group_purchase_product_usage"/>
+            </xpath>
+        </field>
+    </record>
+</odoo>
```
Comment (nikul-serpentcs) [inline purchase_product_usage/views/purchase_order_line_view.xml]: @aheficent add new blank line

File: purchase_product_usage/tests/test_purchase_product_usage.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        usage = cls.env['purchase.product.usage'].create({
+            'name': 'Personal',
+            'code': 'Ps',
+            'account_id': cls.account.id
+        })
+        categ_manual = cls.env['product.category'].create({
+            'name': 'Manual',
+            'property_valuation': 'manual_periodic',
+        })
+        cls.product_1 = cls.env['product.product'].create({
```
Comment (nikul-serpentcs) [inline purchase_product_usage/tests/test_purchase_product_usage.py]: @aheficent `cls.env['product.product']` use multiple time so use as like global variable

**Review by nikul-serpentcs — CHANGES_REQUESTED** [review]
Improve Code

**Comment by AaronHForgeFlow** [comment]
@nikul-serpentcs Thanks for the tip. Done.

File: purchase_product_usage/views/purchase_order_line_view.xml [hunk 3]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <record id="purchase_order_form" model="ir.ui.view">
+        <field name="name">purchase.order.form</field>
+        <field name="model">purchase.order</field>
+        <field name="inherit_id" ref="purchase.purchase_order_form"/>
+        <field name="arch" type="xml">
+            <xpath expr="//field[@name='order_line']/tree/field[@name='name']"
+                   position="after">
+                <field name="usage_id" groups="purchase_product_usage.group_purchase_product_usage"/>
```
Comment (LoisRForgeFlow) [inline purchase_product_usage/views/purchase_order_line_view.xml]: remove the group here, so anyone can specify a usage for a PO line.

File: purchase_product_usage/readme/CONFIGURE.rst [hunk 4]
```diff
@@ -0,0 +1,3 @@
+Activate usages for purchased products on settings
```
Comment (LoisRForgeFlow) [inline purchase_product_usage/readme/CONFIGURE.rst]: drop this line

**Review by LoisRForgeFlow — COMMENTED** [review]
Some comments to align with the fact that you dropped the settings.

**Comment by LoisRForgeFlow** [comment]
I miss the usage in the form view for purchase lines: 
image

**Comment by AaronHForgeFlow** [comment]
Latest changes:

- Add a default product for usages. The product is added automatically to the PO line when selecting the usage.
- Fixes commented by @lreficent:
  - Add usage to PO line form view 
  - Remove groups from view so every purchase user can see and select

**Comment by AaronHForgeFlow** [comment]
I forgot to add the domain in the products. Only consumables or services can be selected.

File: purchase_product_usage/models/purchase_order_line.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+class PurchaseOrderLine(models.Model):
+    _inherit = "purchase.order.line"
+
+    usage_id = fields.Many2one(
+        comodel_name='purchase.product.usage',
+        string="Usage")
+
+    @api.onchange('usage_id')
+    def onchange_usage_id(self):
+        if self.usage_id.product_id:
```
Comment (LoisRForgeFlow) [inline purchase_product_usage/models/purchase_order_line.py]: `... and not self.product_id`

Comment (AaronHForgeFlow) [reply purchase_product_usage/models/purchase_order_line.py]: :sweat_smile:

**Review by LoisRForgeFlow — APPROVED** [review]
Thanks! this looks good for me now :+1:

**Comment by LoisRForgeFlow** [comment]
@aheficent could you squash?

@nikul-serpentcs please update your review

**Comment by AaronHForgeFlow** [comment]
Hi @nikul-serpentcs  @hveficent Could you update your review please?

**Review by nikul-serpentcs — APPROVED** [review]
Code Review LGTM :+1:

**Review by JordiBForgeFlow — APPROVED** [review]
:+1:  Technical and functional review. Tested on runbot.

**Review by JordiBForgeFlow — APPROVED** [review]
:+1: Code review and functional tests. Looks good!

**Pull request merged**